From ae2f116819c34180e3ad69d702f8b6e9cc8ce7a3 Mon Sep 17 00:00:00 2001
From: Ravi Vembu <ravi.v@motorola.com>
Date: Mon, 21 Feb 2011 14:49:21 -0600
Subject: OMAP4: EHCI: Fix kernel panic during shutdown of ehci driver

    When the kernel is built with the ehci driver and USB runtime
    power management enabled, the kernel will panic when the
    ehci_hcd_omap_shutdown() function is called if the ehci driver
    is suspended. Fix this by resuming the bus before shutting down
    the driver.

Change-Id: Iee7f579871f33103a487f6eb593b83f3743636dc
Signed-off-by: Jon Hunter <jon-hunter@ti.com>
Signed-off-by: Ravi Vembu <a18394@motorola.com>
---
 drivers/usb/host/ehci-omap.c |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/drivers/usb/host/ehci-omap.c b/drivers/usb/host/ehci-omap.c
index 79470ad..8cec285 100644
--- a/drivers/usb/host/ehci-omap.c
+++ b/drivers/usb/host/ehci-omap.c
@@ -281,8 +281,13 @@ static void ehci_hcd_omap_shutdown(struct platform_device *pdev)
 {
 	struct ehci_hcd_omap *omap = platform_get_drvdata(pdev);
 	struct usb_hcd *hcd = ehci_to_hcd(omap->ehci);
+	struct uhhtll_apis *uhhtllp = pdev->dev.platform_data;
+	int ret = 0;
+
+	if (uhhtllp && uhhtllp->resume)
+		ret = uhhtllp->resume(OMAP_EHCI);
 
-	if (hcd->driver->shutdown)
+	if (!ret && hcd->driver->shutdown)
 		hcd->driver->shutdown(hcd);
 }
 
-- 
1.6.0.4

